packages/loopring_v3/contracts/iface/IExchangeV3.sol

```diff
@@ -889,84 +778,110 @@ contract IExchangeV3 is IExchange
     ///      Funds from withdrawal requests will be sent to the account owner
-    ///      by the operator in distributeWithdrawals. The user can however
-    ///      choose to withdraw earlier if he wants, or will need to call this
-    ///      manually if nobody calls distributeWithdrawals.
-    ///
-    ///      Funds can only be withdrawn from requests processed in a
-    ///      finalized block (i.e. a block that can never be reverted).
+    ///      immediately by the operator when the block is submitted.
+    ///      The user will however need to call this manually if the transfer failed.
     ///
-    /// @param  blockIdx The block the withdrawal requests were committed in
-    /// @param  slotIdx The index in the list of withdrawals that were processed
-    ///                 by the operator. It is not possible for users to know
-    ///                 what this index will be for their withdrawal request.
+    /// @param  owner The address of the account the withdrawal was done for.
+    /// @param  token The token address
     function withdrawFromApprovedWithdrawal(
-        uint blockIdx,
-        uint slotIdx
+        address owner,
+        address token
         )
         external;
 
-    /// @dev Allows the operator to withdraw the fees he earned by processing the
-    ///      deposit and onchain withdrawal requests.
+    /// @dev Gets the amount that can be withdrawn immediately with `withdrawFromApprovedWithdrawal`.
+    /// @param  owner The address of the account the withdrawal was done for.
+    /// @param  token The token address
+    /// @return The amount withdrawable
+    function getAmountWithdrawable(
+        address owner,
+        address token
+        )
+        external
+        view
+        returns (uint);
+
+    // -- Agents --
+
+    /// @dev Authorizes/Deauthorizes agents for an account.
+    ///      An agent is allowed to authorize onchain operations for the account owner.
+    ///      By definition the account owner is an agent for himself.
     ///
-    ///      This function is only callable by the exchange operator.
+    ///      This function can only be called by an agent.
     ///
-    ///      The block fee can only be withdrawn from finalized blocks
-    ///      (i.e. blocks that can never be reverted).
+    /// @param owner The account owner.
+    /// @param agents The agents to be authorized/deauthorized.
+    /// @param authorized True to authorize the agent, false to deauthorize
+    function authorizeAgents(
+        address   owner,
+        address[] calldata agents,
+        bool[]    calldata authorized
+        )
+        external;
+
+    /// @dev Returns whether an agent address is an agent of an account owner
+    /// @param owner The account owner.
+    /// @param agent The agent address
+    /// @return True if the agent address is an agent for the account owner, else false
+    function isAgent(address owner, address agent)
```

Comment:
I think the UI for the exchange is a better place for this, even if it means having to trust the UI code. I think it will be the only realistic way people will check this.

I don't think we can expect users to go to the contract and get a list of their agents from time to time and see if there are some unexpected addresses. Especially because this will just be a list of Ethereum addresses that will likely contain some actual Agents, but users would need to be able to detect incorrect ones from attacker ones just on the address. That's a lot to expect from them and I don't really see that happening for 99.9% of users.

It's still possible to fetch a list of agents from the contract using events on etherscan for example, so technical people that want to go check the agents without a UI can still do it. We could even allow people to check their agents on Dune if we want so the query is open source.